### PR TITLE
[SPARK-32878][CORE] Avoid scheduling TaskSetManager which has no pending tasks

### PR DESCRIPTION
**Pull request by Ngone51**
### What changes were proposed in this pull request?

This PR proposes to avoid scheduling the (non-zombie) TaskSetManager which has no pending tasks.

### Why are the changes needed?

Currently, Spark always tries to schedule a (non-zombie) TaskSetManager even if it has no pending tasks. This causes notable problems for the barrier TaskSetManager: 1. `calculateAvailableSlots` can be called for multiple times for a launched barrier TaskSetManager; 2. user would see "Skip current round of resource offers for barrier stage" log message for  
a launched barrier TaskSetManager all the time until the barrier TaskSetManager finishes, which is quite confused.

Besides, scheduling a TaskSetManager always involves many function invocations even if there're no pending tasks.

Therefore, I think we can skip those un-schedulable TasksetManagers to avoid the potential overhead.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Pass existing tests.

### CODE REVIEW

**Comment by Ngone51** [comment]
@tgravescs @jiangxb1987 @cloud-fan Any idea on this?

File: core/src/main/scala/org/apache/spark/scheduler/Pool.scala [hunk 1]
```diff
@@ -107,7 +109,7 @@ private[spark] class Pool(
     for (schedulable <- sortedSchedulableQueue) {
       sortedTaskSetQueue ++= schedulable.getSortedTaskSetQueue
     }
-    sortedTaskSetQueue
+    sortedTaskSetQueue.filter(_.isSchedulable)
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/scheduler/Pool.scala:112]: minor nit: Move this check to `schedulable.getSortedTaskSetQueue` in the loop ?

**Review by mridulm — APPROVED** [review]

**Review by tgravescs — APPROVED** [review]
definitely seems like a good thing in general, I was a little worried about other side affects that happen in the scheduling loop but the only one I saw was with locality which doesn't matter since no pending tasks so I think its ok.

**Review by jiangxb1987 — APPROVED** [review]
LGTM

File: core/src/main/scala/org/apache/spark/scheduler/Pool.scala (hunk 1 above)
Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/scheduler/Pool.scala:112]: +1

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. It seems that @mridulm 's comment is the last one we need to do.

**Comment by Ngone51** [comment]
Thanks all! I've addressed the comment.

**Comment by dongjoon-hyun** [comment]
Merged to master for Apache Spark 3.1.0 on December 2020.

**Comment by dongjoon-hyun** [comment]
Thank you, @Ngone51 and all!

**Pull request closed without merging**